drivers/crypto/crypto_si32.c

```diff
@@ -0,0 +1,1157 @@
+
+	switch (channel) {
+	case DMA_CHANNEL_ID_RX:
+		LOG_DBG("AES0 RX DMA channel %s", result);
+		k_sem_give(&crypto_si32_work_done);
+		break;
+	case DMA_CHANNEL_ID_TX:
+		LOG_DBG("AES0 TX DMA channel %s", result);
+		break;
+	case DMA_CHANNEL_ID_XOR:
+		LOG_DBG("AES0 XOR DMA channel %s", result);
+		break;
+	}
+}
+
+static int crypto_si32_query_hw_caps(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	return (CAP_RAW_KEY | CAP_INPLACE_OPS | CAP_SEPARATE_IO_BUFS | CAP_SYNC_OPS |
+		CAP_NO_IV_PREFIX);
+}
+
+static void crypto_si32_irq_error_handler(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	/* 12.3 Interrupts: An AES0 error interrupt can be generated whenever an input/output data
+	 * FIFO overrun (DORF = 1) or underrun (DURF = 1) error occurs, or when an XOR data FIFO
+	 * overrun (XORF = 1) occurs.
+	 */
+	if (SI32_AES_0->STATUS.ERRI) {
+		LOG_ERR("AES0 FIFO overrun (%u), underrun (%u), XOR FIF0 overrun (%u)",
+			SI32_AES_0->STATUS.DORF, SI32_AES_0->STATUS.DURF, SI32_AES_0->STATUS.XORF);
+		SI32_AES_A_clear_error_interrupt(SI32_AES_0);
+	}
+}
+
+/* For simplicity, the AES HW does not get turned of when not in use. */
+static int crypto_si32_init(const struct device *dev)
+{
+	ARG_UNUSED(dev);
+
+	/* Enable clock for AES HW */
+	SI32_CLKCTRL_A_enable_apb_to_modules_0(SI32_CLKCTRL_0, SI32_CLKCTRL_A_APBCLKG0_AES0);
+
+	/* To use the AES0 module, firmware must first clear the RESET bit before initializing the
+	 * registers.
+	 */
+	SI32_AES_A_reset_module(SI32_AES_0);
+
+	__ASSERT(SI32_AES_0->CONTROL.RESET == 0, "Reset done");
+
+	/* 12.3. Interrupts: The completion interrupt should only be used in conjunction
+	 * with software mode (SWMDEN bit is set to 1) and not with DMA operations, where the DMA
+	 * completion interrupt should be used.
+	 */
+	SI32_AES_A_disable_operation_complete_interrupt(SI32_AES_0); /* default */
+
+	/* 12.3. Interrupts: The error interrupt should always be enabled (ERRIEN = 1), even when
+	 * using the DMA with the AES module.
+	 */
+	SI32_AES_A_enable_error_interrupt(SI32_AES_0);
+
+	/* Install error handler */
+	IRQ_CONNECT(DT_INST_IRQN(0), DT_INST_IRQ(0, priority), crypto_si32_irq_error_handler,
+		    DEVICE_DT_INST_GET(0), 0);
+	irq_enable(DT_INST_IRQN(0));
+
+	/* Halt AES0 module on debug breakpoint */
+	SI32_AES_A_enable_stall_in_debug_mode(SI32_AES_0);
+
+	/* For peripheral transfers, firmware should configure the peripheral for the DMA transfer
+	 * and set the device’s DMA crossbar (DMAXBAR) to map a DMA channel to the peripheral.
+	 */
+	SI32_DMAXBAR_A_select_channel_peripheral(SI32_DMAXBAR_0, SI32_DMAXBAR_CHAN5_AES0_TX);
+	SI32_DMAXBAR_A_select_channel_peripheral(SI32_DMAXBAR_0, SI32_DMAXBAR_CHAN6_AES0_RX);
+	SI32_DMAXBAR_A_select_channel_peripheral(SI32_DMAXBAR_0, SI32_DMAXBAR_CHAN7_AES0_XOR);
+
+	return 0;
+}
+
+static int crypto_si32_aes_set_key(const uint8_t *key, uint8_t key_len)
+{
+	switch (key_len) {
+	case 32:
+		SI32_AES_0->HWKEY7.U32 = *((uint32_t *)key + 7);
```

Comment:
Do you think we could get rid of all these casts? I feel `key` could be changed in in `const uint32_t *`.
That's said, it is not a blocker.

Reply:
Changed for `const uint32_t *` and using them as array.